packages/frame/src/stats.ts

```
@@ -1,82 +1,62 @@
-const getQuantitiveStats: GetQuantitiveStats = weakMemoize(frame => {
-  const quantitiveColumns = frame.schema.filter(column => column.type === "number").map(column => column.name);
-  let max: number[] = [];
-  let min: number[] = [];
-
-  frame.forEach(quantitiveColumns, (...values) => {
-    if (max.length === 0) {
-      max = [...values];
-      min = [...values];
-    } else {
-      values.forEach((value, i) => {
-        max[i] = Math.max(value, max[i]);
-        min[i] = Math.min(value, min[i]);
-      });
-    }
-  });
-
-  return {
-    min: zip(quantitiveColumns, min),
-    max: zip(quantitiveColumns, max),
-  };
-});
-
-type GetCategoricalStats = <Name extends string>(frame: IteratableFrame<Name>) => { unique: Record<Name, string[]> };
-
-const getCategoricalStats: GetCategoricalStats = weakMemoize(frame => {
-  const categoricalColumns = frame.schema.filter(column => column.type === "string").map(column => column.name);
-  const unique: Array<Set<string>> = categoricalColumns.map(() => new Set<string>());
+  column: ColumnCursor<Name>,
+): StatsCacheItem => {
+  if (!statsCache.has(frame)) {
+    statsCache.set(frame, {});
+  }
+  const cacheEntry = statsCache.get(frame)!;
+  if (!cacheEntry[column.index]) {
+    cacheEntry[column.index] = {};
+  }
+  return cacheEntry[column.index];
+};
 
-  frame.forEach(categoricalColumns, (...values) => {
-    values.forEach((value, i) => {
-      unique[i].add(value);
+export const maxValue = <Name extends string>(frame: IteratableFrame<Name>, column: ColumnCursor<Name>): number => {
+  const cacheItem = getStatsCacheItem(frame, column);
+  if (cacheItem.max === undefined) {
+    // if (process.env.NODE_ENV === "development") {
+    //   if (frame.schema[column.index].type !== "number") {
+    //     console.warn(`Trying to get max value of none-numeric column ${column.name}`);
+    //   }
+    // }
+    // if (frame.length() === 0) {
```

shall we throw exception for the case with no rows or return undefined?

---

Why `0` and not `100` or `-100`? There is no way to select number if there are no numbers, so I treat as exception situation. We can return `null`, which would represent "no result for this operation" or throw exception. Frame can be empty for sure, but max for empty frame doesn't make sense as a for number devision by 0.

---

This is a great comparison with division by 0: which is an exception. In this case, I think the best thing to do is to `throw` and let userland handle the exception accordingly.

---

> Why 0 and not 100 or -100? There is no way to select number if there are no numbers, so I treat as exception situation.

I saw it more as a reduce operation than a select operation. I thought it would reduce a column to give a user the max? Apparently I missed something. 😋 

> I would return 0 actually, unless I missed something.

---

Out of curiosity what JS think max of empty list
```js
Math.max(1,2,3)
3
Math.max(1,2)
2
Math.max(1)
1
Math.max()
-Infinity
```

We can do anything we want, my main concern, is that you should not ask max of empty list, it means you have an error somewhere in your logic, and returning value instead of throwing error will hide an error.

---

This is not critical. I can create a ticket and handle it later